src/polydatum/dal.py

```diff
@@ -1,20 +1,236 @@
+        # to continue walking the path. Everything after the first missing
+        # service/method will be `(location, None)`.
+        if service_or_method:
+
+            # This if condition is handling the case of the first loop here.
+            # we cannot use attribute access on the dal directly because of how
+            # attribute access is deferred with DalMethodRequest objects.
+            if isinstance(service_or_method, dict):
+
+                # If the code being resolved has typo'd a service name, this
+                # could be returning something that is not a service.
+                service_or_method = service_or_method.get(path_segment.name)
+            else:
+                # second time through (and beyond), service or method is a real
+                # service or method, and we do not need to use a special case for
+                # finding the first service.
+                try:
+                    service_or_method = getattr(service_or_method, path_segment.name)
+                    if not service_or_method and callable(service_or_method):
+                        raise DalMethodError(request, path=tuple(location))
+
+                except KeyError:
+                    raise DalMethodError(request, path=tuple(location))
+        else:
+            raise DalMethodError(request, path=tuple(location))
+    return service_or_method
+
+
+class DalMethodResolverMiddleware(MethodMiddleware):
+    """
+    A Method Middleware that resolves deferred dal attribute access to
+    underlying services/methods that has to be called
+    """
+
+    def __call__(self, request: DalMethodRequest, handler: Callable):
+        """
+        Args:
+             request: Input from caller
+             handler: Downstream middleware or actual DAL method handler
+        """
+        request.dal_method = resolve(request)
+        return handler(request)
+
+
+def default_handle_dal_method(request: DalMethodRequest):
+    """
+    The default method middleware handler for a DalMethodRequester
+
+    Args:
+        request (DalMethodRequest): The method request context.
+
+    Returns: Mixed
+    """
+    assert request.dal_method, "DAL method not resolved"
+    return request.dal_method(*request.args, **request.kwargs)
+
+
+class InvalidMiddleware(Exception):
+    pass
 
 
 class DataAccessLayer(object):
     """
     Gives you access to a DataManager's services.
     """
 
-    def __init__(self, data_manager):
+    # default middleware classes need to be instantiated before being
+    # passed to the init function because the resulting object is called
+    # directly (__call__).
+    def __init__(
+        self,
+        data_manager,
+        middleware=None,
+        default_middleware=(DalMethodResolverMiddleware(),),
+        handler=default_handle_dal_method
+    ):
         self._services = {}
         self._data_manager = data_manager
+        self._handler = handler
+        self._reversed_middleware = []
```

Comment:
Why store them reversed? I'd expect `for m in reversed(self.middleware):`

Reply:
only, only, only for the purpose of testing. The instance attr is not used outside of the __init__, but we wanted a way to explicitly check the order in tests. If we don't save this off (which kinda sucks since it's not used outside of tests), we'd have to dig into the callable chains and that gets crazy.

Reply:
Only need to test input/output of a call that goes through middleware. It doesn't actually matter what order the middleware is stored as. What matters is that the input/output is mutated by middleware in the order you expect.